# Marketing Campaign Analysis 

## 1. Introduction

### Background

There are 3 different account types, small, medium and big hospitals. 
The management team is looking to review the marketing campaigns that have been run to grow in-organic revenue.
This project utilizes linear regression to identify the best marketing channel to drive sales.

### Questions

- How much does each marketing campaign and sales contact impact sales ?
- Does the effectiveness vary between different client types?

### Project Plan

1. Understand the background of the problem and what we are trying to solve
2. Import Data
2. Add required features
3. Explore the data and conduct data analysis
4. Correlation analysis
5. Recommendations and findings

##    2. Import Data

In [10]:
import pandas as pd
import numpy as np
import seaborn as sns
import scipy.stats

In [123]:
# Import dataset and understand it

df = pd.read_csv(r"C:\Users\Benja\Desktop\Projects\Data Analyst Bootcamp\New folder\Campaign-Data.csv")
df.head()

Client ID      Client Type  Number of Customers  Montly Target  Zip Code  \
0  ID-987275  Medium Facility                 2800            125      1003   
1  ID-987275  Medium Facility                 2800            125      1003   
2  ID-987275  Medium Facility                 2800            125      1003   
3  ID-987275  Medium Facility                 2800            125      1003   
4  ID-987275  Medium Facility                 2800            125      1003   

  Calendardate  Amount Collected  Unit Sold  Campaign (Email)  \
0   16-01-2014                 0          0               0.0   
1   16-02-2014           3409460         24               0.0   
2   18-03-2014          10228384         75               0.0   
3   18-04-2014          17047304        123               0.0   
4   19-05-2014          23866224        171               0.0   

   Campaign (Flyer)  Campaign (Phone)  Sales Contact 1  Sales Contact 2  \
0               0.0               0.0              0.0              0.0   
1               0.0               0.0              0.0              0.0   
2               0.0               0.0              0.0              0.0   
3               0.0               0.0              0.0        3547500.0   
4               0.0               0.0              0.0              0.0   

   Sales Contact 3  Sales Contact 4  Sales Contact 5 Number of Competition  
0              0.0              0.0              0.0                   Low  
1              0.0              0.0         322500.0                   Low  
2              0.0              0.0              0.0                   Low  
3        1290000.0              0.0              0.0                   Low  
4              0.0              0.0              0.0                   Low

##    3. Feature Engineering

In [ ]:
# Create new variables for Year and Month

df['Calendardate'] = pd.to_datetime(df['Calendardate']) # Convert Calendardate to a DateTime object
df['Calendar_Month'] = df['Calendardate'].dt.month
df['Calendar_Year'] = df['Calendardate'].dt.year

##    4. Exploratory Data Analysis and Statistical Analysis

### Goals

- Understand the distribution of data across diferent client types
- Understand the difference in sales between each client type
- Answer how marketing campaign impacts sales (using correlation & regression)
- Answer how competition impacts sales
- Answer how different types of client should have different marketing campaigns and strategies

### Data distribution

In [14]:
df['Client Type'].value_counts(normalize=True) # Normalize=True allows for calculation of proportion instead of unique counts

Large Facility      0.459677
Small Facility      0.282258
Medium Facility     0.169355
Private Facility    0.088710
Name: Client Type, dtype: float64

In [16]:
pd.crosstab(df['Number of Competition'],df['Client Type'],margins=True,normalize='columns')

Client Type            Large Facility   Medium Facility  Private Facility  \
Number of Competition                                                       
High                          0.166667         0.166667          0.166667   
Low                           0.833333         0.833333          0.833333   

Client Type            Small Facility        All  
Number of Competition                             
High                          0.166667  0.166667  
Low                           0.833333  0.833333

In [125]:
df.groupby('Number of Competition').mean()

Interesting that the Medium Facility has a higher average Amount Collected and Number of Customers compared to the Large Facility

In [128]:
df.groupby('Client Type').mean()

Number of Customers  Montly Target  Zip Code  \
Client Type                                                      
Large Facility            1380.842105      71.578947    1003.0   
Medium Facility           3940.761905     202.857143    1003.0   
Private Facility           400.727273      20.454545    1003.0   
Small Facility             422.514286      21.285714    1003.0   

                  Amount Collected   Unit Sold  Campaign (Email)  \
Client Type                                                        
Large Facility        1.999880e+07  143.098684     142273.609649   
Medium Facility       4.075997e+07  290.583333     437217.097817   
Private Facility      5.030246e+06   35.784091       5183.715152   
Small Facility        1.637759e+06   11.689286      11975.986310   

                  Campaign (Flyer)  Campaign (Phone)  Sales Contact 1  \
Client Type                                                             
Large Facility        8.192056e+05      45595.436623    133667.763158   
Medium Facility       1.552603e+06      49176.847619    398645.833333   
Private Facility      2.272919e+05       5522.470455      1221.590909   
Small Facility        9.120875e+04          0.000000      8062.500000   

                  Sales Contact 2  Sales Contact 3  Sales Contact 4  \
Client Type                                                           
Large Facility       2.034013e+06     2.017039e+06    119287.280702   
Medium Facility      4.822783e+06     4.698646e+06     85104.166667   
Private Facility     6.376705e+05     4.434375e+05      3664.772727   
Small Facility       7.617143e+05     3.727946e+05      4223.214286   

                  Sales Contact 5  
Client Type                        
Large Facility       16266.447368  
Medium Facility      33273.809524  
Private Facility     12215.909091  
Small Facility        1535.714286

In [130]:
# Correlation between Amount Collected and other variables

df.corr()[['Amount Collected']] 


Amount Collected
Number of Customers          0.607496
Montly Target                0.608204
Zip Code                          NaN
Amount Collected             1.000000
Unit Sold                    0.997515
Campaign (Email)             0.248235
Campaign (Flyer)             0.444337
Campaign (Phone)             0.034858
Sales Contact 1              0.277478
Sales Contact 2              0.552112
Sales Contact 3              0.357887
Sales Contact 4              0.236165
Sales Contact 5              0.095795

##    5. Correlation Analysis

In [43]:
# Correlation of each marketing strategy with sales (amount collected)

import seaborn as sns
cm = sns.light_palette("green",as_cmap=True)
correlation_analysis = pd.DataFrame(df[['Amount Collected','Campaign (Email)','Campaign (Flyer)','Campaign (Phone)','Sales Contact 1','Sales Contact 2','Sales Contact 3','Sales Contact 4','Sales Contact 5']].corr()['Amount Collected']).reset_index()
correlation_analysis.columns = ['Variable','Degree of Linear Impact (Correlation)']
correlation_analysis=correlation_analysis.sort_values('Degree of Linear Impact (Correlation)',ascending=False)
correlation_analysis=correlation_analysis[correlation_analysis['Variable']!='Amount Collected'].reset_index(drop=True)
correlation_analysis.style.background_gradient(cmap=cm).format(precision=2)

In [57]:
# Ordered by the account type and in descending order

import seaborn as sns
cm = sns.light_palette("purple",as_cmap=True)
correlation_analysis = pd.DataFrame(df.groupby('Client Type')[['Amount Collected','Campaign (Email)','Campaign (Flyer)'
        ,'Campaign (Phone)','Sales Contact 1','Sales Contact 2','Sales Contact 3','Sales Contact 4'
        ,'Sales Contact 5']].corr()['Amount Collected']).reset_index()
correlation_analysis=correlation_analysis.sort_values(['Client Type','Amount Collected'],ascending=False)
correlation_analysis.columns = ['Account Type','Variable','Degree of Linear Impact (Correlation)']
correlation_analysis=correlation_analysis[correlation_analysis['Variable']!='Amount Collected'].reset_index(drop=True)
correlation_analysis.style.background_gradient(cmap=cm).format(precision=2)

In [76]:
# Linear regression model

import statsmodels.formula.api as smf

for column in df.columns:
    new_column_name = column.replace(" ", "_").replace("(", "").replace(")", "")
    df.rename(columns={column: new_column_name}, inplace=True)  # Assign the modified column name back to the DataFrame


    
results = smf.ols('Amount_Collected ~ Campaign_Email+Campaign_Flyer+Campaign_Phone+\
        Sales_Contact_1+Sales_Contact_2+Sales_Contact_3+Sales_Contact_4+Sales_Contact_5',data=df).fit()
print(results.summary().tables[1])

                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept        1.481e+06   5.12e+05      2.891      0.004    4.77e+05    2.49e+06
Campaign_Email      0.7932      0.597      1.329      0.184      -0.377       1.963
Campaign_Flyer      3.3376      0.260     12.831      0.000       2.828       3.848
Campaign_Phone      0.0734      1.053      0.070      0.944      -1.991       2.137
Sales_Contact_1     4.2368      0.415     10.207      0.000       3.423       5.051
Sales_Contact_2     3.6382      0.129     28.155      0.000       3.385       3.892
Sales_Contact_3     2.3432      0.131     17.925      0.000       2.087       2.600
Sales_Contact_4    10.9478      1.060     10.331      0.000       8.870      13.026
Sales_Contact_5     3.5078      4.549      0.771      0.441      -5.412      12.428


In [66]:
# Marketing strategies which have statistically significant relationships and their impact (coef)

df_results = pd.read_html(results.summary().tables[1].as_html(),header=0,index_col=0)[0]
df_results = df_results.reset_index()
df_results = df_results[df_results['P>|t|']<0.05][['index','coef']] # P-value < 0.05 indicates a statistically significant relationship
df_results

index          coef
0        Intercept  1.481000e+06
2   Campaign_Flyer  3.337600e+00
4  Sales_Contact_1  4.236800e+00
5  Sales_Contact_2  3.638200e+00
6  Sales_Contact_3  2.343200e+00
7  Sales_Contact_4  1.094780e+01

In [116]:
# Saving results into a new dataframe

consolidated_summary = pd.DataFrame()

for account_type in list(set(list(df['Client_Type']))): # For each account type eg. small/med/large facility
    temp_df = df[df['Client_Type'] == account_type].copy()
    results = smf.ols('Amount_Collected ~ Campaign_Email+Campaign_Flyer+Campaign_Phone+\
            Sales_Contact_1+Sales_Contact_2+Sales_Contact_3+Sales_Contact_4+Sales_Contact_5',data=temp_df).fit()
    df_results2 = pd.read_html(results.summary().tables[1].as_html(),header=0,index_col=0)[0].reset_index() # Table 1 contains coef,std-err,t-values and p-values
    df_results2 = df_results2[df_results2['P>|t|']<0.05][['index','coef']] # Only showing varaibles that pass significance test
    df_results2.columns = ['Variable','Coefficient']
    df_results2['Account Type'] = account_type
    df_results2 = df_results2[df_results2['Variable']!='Intercept'] # Excluding intercept values
    df['Account Type'] = account_type
    df_results2 = df_results2.sort_values('Coefficient',ascending = False) # Sort results by coefficient from highest to lowest
    print(account_type)
    consolidated_summary = consolidated_summary.append(df_results2)
    print(df_results2)
    

{'Private Facility', 'Medium Facility', 'Large Facility ', 'Small Facility '}
Private Facility
          Variable  Coefficient      Account Type
5  Sales_Contact_2       6.6223  Private Facility
Medium Facility
          Variable  Coefficient     Account Type
2   Campaign_Flyer       4.1059  Medium Facility
5  Sales_Contact_2       3.5778  Medium Facility
4  Sales_Contact_1       3.1365  Medium Facility
6  Sales_Contact_3       2.1174  Medium Facility
Large Facility 
          Variable  Coefficient     Account Type
4  Sales_Contact_1      11.6731  Large Facility 
7  Sales_Contact_4      10.6145  Large Facility 
5  Sales_Contact_2       4.0031  Large Facility 
2   Campaign_Flyer       2.7204  Large Facility 
6  Sales_Contact_3       2.0316  Large Facility 
3   Campaign_Phone      -3.5361  Large Facility 
Small Facility 
          Variable   Coefficient     Account Type
5  Sales_Contact_2  8.101000e-01  Small Facility 
3   Campaign_Phone  9.753000e-07  Small Facility 


C:\Users\Benja\AppData\Local\Temp\ipykernel_15868\1980759352.py:18: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  consolidated_summary = consolidated_summary.append(df_results2)
C:\Users\Benja\AppData\Local\Temp\ipykernel_15868\1980759352.py:18: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  consolidated_summary = consolidated_summary.append(df_results2)
C:\Users\Benja\AppData\Local\Temp\ipykernel_15868\1980759352.py:18: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  consolidated_summary = consolidated_summary.append(df_results2)
C:\Users\Benja\AppData\Local\Temp\ipykernel_15868\1980759352.py:18: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead

##    6. Conclusion

The ROI in the table below tells us the return derived from each dollar spent. 
The results show that the different sales contacts and campaigns perform with varying effectivenesss within the group of accounts.

The Large Facility shows excellent results with sales contact 1 & 4 with an average ROI of $ 11.67 and $ 10.61 respectively. 
Flyer campaigns are doing decently well with ROI $ 2.72 while it looks like it makes financial sense to drop phone campaigns which seems to be working against the company's sales.

In [117]:
consolidated_summary

Variable   Coefficient      Account Type
5  Sales_Contact_2  6.622300e+00  Private Facility
2   Campaign_Flyer  4.105900e+00   Medium Facility
5  Sales_Contact_2  3.577800e+00   Medium Facility
4  Sales_Contact_1  3.136500e+00   Medium Facility
6  Sales_Contact_3  2.117400e+00   Medium Facility
4  Sales_Contact_1  1.167310e+01   Large Facility 
7  Sales_Contact_4  1.061450e+01   Large Facility 
5  Sales_Contact_2  4.003100e+00   Large Facility 
2   Campaign_Flyer  2.720400e+00   Large Facility 
6  Sales_Contact_3  2.031600e+00   Large Facility 
3   Campaign_Phone -3.536100e+00   Large Facility 
5  Sales_Contact_2  8.101000e-01   Small Facility 
3   Campaign_Phone  9.753000e-07   Small Facility

In [118]:
consolidated_summary.reset_index(inplace=True)
consolidated_summary.drop('index',inplace=True,axis=1)
consolidated_summary.columns = ['Variable','ROI','Account Type']
#consolidated_summary['ROI'] = consolidated_summary['ROI'].map(lambda x: "${:.2f}".format(x))
consolidated_summary.style.background_gradient(cmap='RdYlGn')